reps/2022-04-21-object_store_HA.md

```diff
@@ -0,0 +1,197 @@
+## Summary - Object Store High Availability
+
+### General Motivation
+
+Ray is a general-purpose and powerful computing framework. With the Object Store, it can be easily extended into a data service to provide data for various distributed tasks. Ray uses decentralized object ownership to avoid centralized bottlenecks in managing an object’s metadata (mainly its reference count and object directory entries), but it difficult to handle failover.
+
+For now, users can only rely on lineage to recover the unavailable object. But lineage has many restriction:
+- Can not recover the object which put in object store via `ray.put`.
+- Can not recover the object returned by actor task.
+- Require task is idempotent.
+
+#### Goal
+
+1. Objects can be specified for high availability mode, other objects are the same as before.
+2. Any high availability objects should still be accessible if we encounter a single-node failure.
+
+### Should this change be within `ray` or outside?
+
+Changes are within Ray core.
+
+## Stewardship
+
+### Required Reviewers
+
+@stephanie-wang, @ericl, @scv119, @kfstorm, @raulchen
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@ericl, @raulchen, @stephanie-wang
+
+## Design and Architecture
+
+### Problem statement
+
+**TODO: Is it better to have a separate section or separate document for the detailed design and let this REP mainly focus on motivation and high-level design?**
+
+#### Problem 1: Object Owner Failure
+
+The owner of an object stores the metadata of the object, such as reference count and locations of the object. If the owner dies, other workers which hold the object ref cannot access the data of the object anymore because all copies of the object will be deleted from the Object Store.
+
+#### Problem 2: Object Borrower Failure
+
+In the chained object borrowing case, the owner of an object is not aware of the indirect borrowers. If all direct borrower fails, the owner will consider the object out-of-scope and GC the object. Accessing the object on indirect borrowers will fail with an `ObjectLostError`.
+
+more details: [issues 18456](https://github.com/ray-project/ray/issues/18456)
+
+#### Problem 3: Loss of All Copies
+
+Data of objects stored in the plasma store. For now, the plasma store is a thread of the raylet process, failure of the raylet process will lose data which store in plasma. Some objects which only one copy in that failed plasma store,  will be unavailable.
+
+### Proposed Design
+
+#### How to solve the object owner failure problem?
+
+**Global Owner**: A group of special Ray actors with `max_restarts=-1` will be created to own all objects of a job created by `ray.put` and the returned object (either from normal-task or actor-task). When one of the special actors fails, the restarted instance will rebuild the reference table based on information provided by Raylets and other non-special workers.
```

Comment:
> A group of special Ray actors

I'm interested in what this means. Is it a new type of core worker (it doesn't seem like it makes sense to do this at the application level)? or should this type of logic go in GCS?

Reply:
I think it is a new type of CoreWorker(inherited from CoreWorker), add some rebuild logic when it restarts, etc.  In short, it will not be much different from the original CoreWorker.